src/Microsoft.ML.Dnn/TensorTypeExtensions.cs

```diff
@@ -3,49 +3,477 @@
+        {
+            unsafe
+            {
+                switch (tensor.dtype)
+                {
+#if _REGEN
+                    %foreach supported_numericals_TF_DataType,supported_numericals,supported_numericals_lowercase%
+                    case TF_DataType.#1:
+                        dst = ((IConvertible) (*(#3*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    %
+#else
+                    case TF_DataType.TF_BOOL:
+                        dst = ((IConvertible) (*(bool*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_UINT8:
+                        dst = ((IConvertible) (*(byte*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_INT16:
+                        dst = ((IConvertible) (*(short*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_UINT16:
+                        dst = ((IConvertible) (*(ushort*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_INT32:
+                        dst = ((IConvertible) (*(int*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_UINT32:
+                        dst = ((IConvertible) (*(uint*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_INT64:
+                        dst = ((IConvertible) (*(long*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_UINT64:
+                        dst = ((IConvertible) (*(ulong*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_DOUBLE:
+                        dst = ((IConvertible) (*(double*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+                    case TF_DataType.TF_FLOAT:
+                        dst = ((IConvertible) (*(float*) tensor.buffer)).ToString(CultureInfo.InvariantCulture);
+                        return;
+#endif
+                    case TF_DataType.TF_STRING:
+                        dst = tensor.StringData()[0];
+                        return;
+                    case TF_DataType.TF_COMPLEX64:
+                    case TF_DataType.TF_COMPLEX128:
+                    default:
+                        throw new NotSupportedException();
+                }
+            }
         }
 
-        public static void CopyTo<T>(this Tensor tensor, Span<T> values) where T: unmanaged
+        public static void CopyTo(this Tensor tensor, Span<string> destination)
```

Comment:
Who is calling this method? We shouldn't have dead code in the repo.

Reply:
Right, `TensorTypeExtensions` is internal. My bad.

Reply:
Eric, I'm curious if this method is for string tensors? If it is we should keep it as it is one of the types we may need to use when extracting output tensors. 

---
In reply to: [331242079](https://github.com/dotnet/machinelearning/pull/4290#discussion_r331242079) [](ancestors = 331242079)

Reply:
If/when we have a caller we can add that support.

As it exists, we could never use this method in ML.NET because we never have `Span<string>` variables. We always code to `ReadOnlyMemory<char>`.

Reply:
In tensorflow, string are stored encoded (usually UTF8) with a header of `ulong length`.
If you are willing to provide me with an example/explanation towards how strings are handled inside ML.NET, I'm willing to write a conversion function from Tensor to ReadOnlyMemory<char> or what ever you use.